### ISSUE TITLE
NoClassDefFoundError in Android 5-6 when built in Android Studio 4.2.1

### ISSUE THREAD

**Issue by aloz77** [hivemq/hivemq-mqtt-client#488]
## Expected behavior

MQTT Client can init and connect with Android 5-6 devices

## Actual behavior

The client is crashing in Android 5 and Android 6 with

```
java.lang.NoClassDefFoundError: com.hivemq.client.internal.mqtt.-$$Lambda$QT9XFB_2yV951PYVP80Cqx3qU9Q
	at com.hivemq.client.internal.mqtt.MqttRxClientBuilderBase.automaticReconnect(MqttRxClientBuilderBase.java:156)
	at com.hivemq.client.internal.mqtt.mqtt3.Mqtt3RxClientViewBuilder.automaticReconnect(Mqtt3RxClientViewBuilder.java:38)
	at com.mypackage.MqttHiveClient.<init>(MqttHiveClient.java:54)
```

No issues with newer Android versions or when built in Android Studio 4.1.x

## To Reproduce

Run the app with following code

### Reproducer code

```
                Mqtt3ClientBuilder mqtt3ClientBuilder = Mqtt3Client.builder()
                        .identifier(identifier)
                        .serverHost(host)
                        .serverPort(port)
                        .simpleAuth(Mqtt3SimpleAuth.builder()
                                .username(username)
                                .password(password.getBytes())
                                .build())
                        .automaticReconnect()
                        .initialDelay(10, TimeUnit.SECONDS)
                        .maxDelay(60, TimeUnit.SECONDS)
                        .applyAutomaticReconnect();
```

dependencies:
```
    implementation 'com.hivemq:hivemq-mqtt-client:1.2.2'
    implementation 'com.hivemq:hivemq-mqtt-client-websocket:1.2.2'
    implementation 'net.sourceforge.streamsupport:android-retrostreams:1.7.3'
    implementation 'net.sourceforge.streamsupport:android-retrofuture:1.7.3'
```

buildscript/dependencies:
```
        classpath 'com.android.tools.build:gradle:4.2.1'
        classpath 'com.github.dcendents:android-maven-gradle-plugin:1.4.1'
        classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.8.5'
        classpath 'gradle.plugin.com.github.sgtsilvio.gradle:android-retrofix:0.3.6'
```

`apply plugin: 'com.github.sgtsilvio.gradle.android-retrofix'`

**Comment by SgtSilvio** [comment]
Hi @aloz77 
did you set the Java version to 1.8, so that lambdas are properly desugared?
```
android {
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
}
```
(https://hivemq.github.io/hivemq-mqtt-client/docs/installation/android/)

**Comment by aloz77** [comment]
Yes, of cause, `JavaVersion.VERSION_1_8` is enabled as in your snippet above.

The issue was introduced by Android 4.2 IMHO.

**Comment by SgtSilvio** [comment]
Did you set?
```
android {
    compileOptions {
        coreLibraryDesugaringEnabled = true
    }
}
```
If not, what is the default value? Maybe it changed in the new version of the Android plugin.

I believe that this setting is not compatible with the android-retrofix plugin.

**Comment by aloz77** [comment]
There was no `coreLibraryDesugaringEnabled` entry in my build.gradle. I tried `coreLibraryDesugaringEnabled false` but this didn't help either.

**Comment by aloz77** [comment]
I tried to build with the old Gradle and Gradle Plugin and it works for Android 5 then. 

```
classpath 'com.android.tools.build:gradle:4.1.3'

distributionUrl=https\://services.gradle.org/distributions/gradle-6.5-bin.zip
```

A fix for the latest Android Studio 4.2.1 would be great.

**Comment by digrec** [comment]
I have exactly the same issue, as aloz77 does, on:
- Android Studio 4.2.1, AGP 4.2.1, on Android 6 device
- Android 7.1.1 device is not affected

```
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
```

```
java.lang.NoClassDefFoundError: com.hivemq.client.internal.mqtt.-$$Lambda$Df8yRNpuHjFWi4vVWQkOSvVrarE
    at com.hivemq.client.internal.mqtt.MqttClientTransportConfigImplBuilder.webSocketConfig(MqttClientTransportConfigImplBuilder.java:206)
    at com.hivemq.client.internal.mqtt.MqttRxClientBuilder.webSocketConfig(MqttRxClientBuilder.java:39)
```
The only workaround for me until now was downgrading AGP to 4.1.3. Maybe it has something to do with [AGP 4.2.1 using Java 8 language level by default](https://developer.android.com/studio/releases/gradle-plugin#java-8-default)?

Disabling core lib desugaring doesn't help.

But enabling it, [like this](https://developer.android.com/studio/write/java8-support.html#library-desugaring) does help:
```
android {
    compileOptions {
        coreLibraryDesugaringEnabled true
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
}

dependencies {
    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.0.9'
    ...
}
```
and the `NoClassDefFoundError` is gone on Android 6.

Is this now recommended solution when using HiveMQ client and AGP 4.2.1, even if I'm using only Kotlin in my project? Or RetroFix plugin needs a fix?

**Comment by SgtSilvio** [comment]
Hi @digrec 
The problem with `coreLibraryDesugaring` is, that Android still only supports a subset of the Java 8 libraries, for example a CompletableFuture backport is missing iirc.
Imho the android-retrofix plugin needs an update to support the new android plugin which seems to have broken backwards compatibility.

**Comment by volkansahin45** [comment]
Same problem. Downgrade AGP 4.1.3 fixed for now.

**Comment by george-theocharis** [comment]
Any progress on this maybe? It doesn't actually work in AGP 7 too. I opened a retrofix issue too before seeing this issue here.

**Comment by codefluencer** [comment]
Having the same problem, any other fixes apart from downgrading gradle ?

**Comment by HiveMQSupport** [comment]
Tune in to our special edition webinar to get first-hand information on the upcoming 4.7 release.

Last chance to register for our special edition webinar on

Wednesday, 15. Sept. 2021 at 5 PM CEST/ 11 AM ET / 8 AM PT

Reserve Your Spot Now
(https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VW20ys7FKvZ8W83b-xZ27rP0nW5g6HvJ4xrVwrN5q1hj95mVy7V3Zsc37CgRfRN4MtXhtkTdVxW3-SW0V1mCmLjW6fYxss3RDfh3VvNctF6cMhMGW5r39yT4lwKsjW3zwhll4m_cnBW69xLfh5YlxkmW4zz9Qd3B3WTLW8vKD-m71bw6sW6vckS429zv-4N8CZXW6TfPK0W6rhr1r8k2zmHVv2sjm7m03jxW2WWxDY2VXGvfVWYtpm306sT8V_L1-V5g55F4W7nJ6TG3qQ9qkW1D9DRh1XTz22W8CF5m223BLlYW5sS3-w6L-njjW3y2F6c1531ywW90CK_-5RWdVTW5gBg2Q3JmFzlVmzrk19cDt3hW7kjWCM6z9HX_W8-22pL6WhqhkW3gkP8j8qYZsPW2sQ9RB2JHD3-W2RMy-16dKGbbW7F8cYp7bRv5SN4m_6Y-K7Yv2VBS0Zs6YYrpGN3jNn1f-g9PSW1XFzmQ4n6cTd3hh81 )

Hi hivemq/hivemq-mqtt-client,

Checking in one last time before our webinar this Wednesday on 'What's New in HiveMQ: Inside the Upcoming HiveMQ 4.7 Release.'

Why You Should Attend:

HiveMQ 4.7 will be released soon. In this webinar, Dominik Obermaier, CTO and co-founder of HiveMQ, and Georg Held, Engineering Manager at HiveMQ, will showcase the new features of HiveMQ 4.7 and discuss how you can benefit from upgrading to the new release.

Some of the highlights of HiveMQ 4.7 include:

- Easier operations for cloud-native deployments

- Exciting new Enterprise Security features

- Next level upgrades to HiveMQ Swarm – the MQTT load testing tool

Register Now
(https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VW20ys7FKvZ8W83b-xZ27rP0nW5g6HvJ4xrVwrN5q1hj95mVy7V3Zsc37CgC60VXCggP8cl0x_W4RN51q6MmqvqW4Fcgy54XfbtdW6DTDVL12qjMLN7k2mrVzLyFtVJLXWL3bPVMTW11hWdS9lKVlCW6XyKd-7QnYdPVvXKfK4w0M-zW3Jp-QN7lrGYjW2s17xY6XbgNSW8-2fvG1ksVMlW5R7lq_6pW8GrW7-9wYg7M1_nRN33qRQQzGyPZW78nFrj6F_f5zW8q9BNL6PM4pxW2SjqLY30gXmnW8t8tg97T7ldVN2qglxfJwsY8W3L3gkM2yt6k_N26BmVNl6dq4W3NQK2Q3KJvB1N46lZ8yxVsZQW99FnVN3dDQtDW3dFW4N13-MzDVvqM0q537BMrW8q-MsS8LqPpVVW3yLM3LxDXbV5DmlT4l4-TQW92vjP92shnsKW56214h6vNpWdW2hZWTc6bSVn4W7f6QHs7D2jsS3l0k1 )

We look forward to seeing you at this interactive session.

Warm Regards,

The HiveMQ Team

HiveMQ GmbH, Ergoldinger Straße 2a, Landshut, Germany

Manage preferences (https://hs-5867930.s.hubspotemail.net/hs/manage-preferences/unsubscribe?languagePreference=en&amp%3Bd=VnzSxg5pR2lTW2p76LK3-1rCSW45Qvcz1ZvHG4W1ZmzRj43X0srW47lLLz4cj-XnW3__Jd63KbsvFW3R0gxR3z8p7cW4hlMkh45RlhdW3P28X24hCPvhW43T2PJ8-zKhjW15gBJz5hrmD_W69gj5n5l4csCW4WlGNj7_QNhYVWv-0T5JhSlvMXk_7Qbw2JJN266kYN3-c9CW58d5Mx4NlYtv34Pr2&amp%3Bv=3&_hsenc=p2ANqtz-8pJL41G9g6ZuDPEZ8TXScloH5dC1QIBUPNHzs77GWGZdBmGGfOjZ6LlV7b5TQPkEiKhDuGnwHIrD4qhn2aN8-GaV1WuYwpm9BPddeVVm-ELqWd_anBEvUpvABhBC-5EXwQzg5ozwPXF-znRIhNEdfCcsVLdQ&_hsmi=159067984 )

You are receiving this email because you filled out a form on our website

or visited us at our booth at a tradeshow.

Twitter (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VW20ys7FKvZ8W83b-xZ27rP0nW5g6HvJ4xrVwrN5q1hkG3pl3rV1-WJV7CgQsbW55wXcB5F7FNvV9K6pD3dpNjsN3xR6nrFm_xmV6rm8Q2lQzqsW9dMmzs30Lg1NN6fHJ0TFWcMpW1Q57Bd4YvBhTW8bcmj_8vGV-2W8qFMmT6PsPR0VtGmdV9c5DP_W1pLB3y6hQctzVXlf1l7X2QywW1w8lYl5bdV1MMsX4gp1GMs9VWrz_D8yh4cxVCY2Td5YH-_mW1zsP_L23wX47VmL-MK1B4zbsW4CWrrX7_5-B_W7F1Qgs5kDZ9rW7hbDql6zgPP2W6t3f286pGwh5W3CPwkY7vMMHCW4j7cg464CS6tW2xT9BW8JHxD4N6w3H4P4W2h6W3LpBMs7B9vlNW5hLGrB92JDP239FR1 )

LinkedIn (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VW20ys7FKvZ8W83b-xZ27rP0nW5g6HvJ4xrVwrN5q1hkZ3pl3LV1-WJV7CgV1JN1V7H94ySN19W6XtmFq5NRrbcW9gTR-93C5bSyW49k8R53TK9bRW16x5wG36PXJLW2Lt2HP9gmn1BN33-b4FcJ4yLW4Ptdkg6qcSfzW3CHQwM4XzJbjW8jzhRT4Mdxm_W1ZRl4N1g4wmdN42RXxVYQBFNW5wf0G645nPsJW4CY8Rf15YRQtM_XSH1SjRz4W31zWb78S87R7N7DSc_qKCZvvN75lKrbRYGqRW7X8TPg27xzZqW3lmpbD2f881BW2Zwd_34cDl9nW90db2Q5q486FW7nn6PL2JHt3-W1-vnhk3L15ZVW59hx9b4DNJ6jW8TR5y53TP26VW3Th-bt4NTQW-W8RZqTg2V5mdHW5PxW5L34hxX-V2nWng9djlrx3g6n1 )

Facebook (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VW20ys7FKvZ8W83b-xZ27rP0nW5g6HvJ4xrVwrN5q1hkG3pl3rV1-WJV7CgNY3W574XMs2KrLkTW3V0r-L2XYbTtW69Lf-m2bqjljW2-Dvnj3VFf5cW6zsnlZ6mVw6MW5f8CSS73D5dtN38cjB7YxmJyW3MqjsH4zyvsPW44FfH_24rMQMW8WNm8q1TZ2mfVFf-Nk1B4GyQN55pkf1mcYsSVyz50V6HDbZsW6K0YlN2WG8LGW64Hvcv4vHXZnW3F0cFL6fGMbtW98BDGb6gp3hBW1VJRLx8SwGczN5tJtGLF_VF5W1yZBWk8gz46_W8XMywj89xFpNW6Sg_M_4m4k1nW6jnLhR45MbdbW6RQ3YS2C7fxxVk1B8s3-5SVgW5Z7c9f7C1HPcW7q9by74P-CTtMR4zK_7yLL13hcW1 )

YouTube (https://cJgkn04.na1.hubspotlinks.com/Btc/LX+113/cJgkn04/VW20ys7FKvZ8W83b-xZ27rP0nW5g6HvJ4xrVwrN5q1hhV5mVxSV3Zsc37CgCLbW3gwT935dy29kW3fnRQk2tQBMPW2qwgpJ4k0g0PN8Zxn5Q4yBwrW1w6Q__29hBdMW7bb4T_67K18SW7r2ZW44CP1w8VPKWQM8WMDgLW9kHbRz7WbMHWW4pg1cH8xZ3g0W76ppWc7cPrGdW836qSw847TxDW6Tzc348_M-5rW1_vgWX3X3ZD4ML79_VL2ZL0W2gmVv131H8HsW4QJcz54rRH_LW78xS9k8bGLmJW6LQxPL6vn326W4Y71yM8vyNDPW1qgK0g5CcqBgW4d2YyH4jYBCVVCNPhR3Sr8sHW8m69Cx4VxRfZN4KBHYTCsQXsW1KNVWp9fjJ_pW25QfGz65Zr1sW2jpzg_3-XsZJW4Zph1H3jd0VZW6ZbmjZ1qN0tZW1vS-GG3WJr2nW4G9tvQ5R-xnY35b51 )

**Comment by SgtSilvio** [comment]
@aloz77 @digrec @volkansahin45 @george-theocharis @codefluencer @HarisHoulis 
the new 0.4.0 release of android-retroffix (https://github.com/SgtSilvio/android-retrofix/releases/tag/v0.4.0) should work with Android Gradle Plugin 4.x and 7.x
Please comment if it works now.

**Comment by aloz77** [comment]
Yes, looks like working now. Tested with Android 6.

**Comment by digrec** [comment]
@SgtSilvio thanks for the fix, the original issue is now resolved. Tested on Android 6 and 7.

But one of our android apps is now failing to build with: `Cause: module-info class is frozen`

I reported the issue here https://github.com/SgtSilvio/android-retrofix/issues/11

Any suggestions?

**Comment by SgtSilvio** [comment]
@digrec should be fixed in the new version 0.4.1 (https://github.com/SgtSilvio/android-retrofix/releases/tag/v0.4.1)
